### ISSUE TITLE
Editor: track posts/pages from block editor vs classic editor

### ISSUE THREAD

**Issue by creativecoder** [Automattic/wp-calypso#45525]
## Requirements

- Add a Tracks event that shows which editor (classic or block) is used to publish a post, page, or custom post type.

## Notes

Reference for related events: pa1C6h-O0-p2

**Comment by pablinos** [comment]
After quite a bit of investigation, I think the best way to do this would be to add something to the API endpoints that would allow us to track new, and updated, posts from Gutenberg. We add `_gutenberg_nonce` to these requests on WordPress.com, so we should be able to filter out Gutenberg API requests from any others.

On Atomic, we would have to handle things differently. In this case, I think we would need to check if the classic editor plugin is activated at the time, as that's the only way for the classic editor to be used rather than the choice that is given on simple sites.

In both situations, I think we may need to store something against the post as the transition in post status can happen asynchronously. The only feasible place for this is in `postmeta`, which we could clear up after the event is fired. It doesn't seem great to be adding metadata for the purposes of analytics, but I'm not sure there's a better approach.

**Comment by creativecoder** [comment]
> I think we would need to check if the classic editor plugin is activated at the time

User's can still publish with either editor when the classic plugin is active (depending on plugin settings), how would we handle differentiating which editor is used?

---

This is sounding very complicated. I'm leaning toward using client side events in each editor instead. It wouldn't line up exactly with wpcom publish events, but I think we'd still be able to gauge classic editor usage in wp-admin and classic block usage in the block editor. What do you think?

**Comment by pablinos** [comment]
> This is sounding very complicated.

Yeah it is, surprisingly so! We're essentially trying to know which editor is loaded again, but this time we're looking to see in the past whether a particular editor was loaded when an action happened.

> User's can still publish with either editor when the classic plugin is active

On simple sites that's the case, as we need to give them that option (essentially the plugin is active for all), but not on Atomic/Jetpack/self hosted sites. Once the classic editor plugin is active all editing is done in the classic editor, or so it seems on my test sites.

> I'm leaning toward using client side events in each editor instead. 

I'm not sure how much easier this would be. We've got to add click events to the publish button in the classic editor (potentially for Jetpack sites too?}, and it's a different stat to the ones we're tracking at the moment, so it's likely to create confusion. 

We could add events to the publish/update button in Gutenberg, with a property of `has_blocks` and then we'd know if the post was a single classic block or not, but I'm not sure how useful that is.

Given this will be more work, I'm starting to question the value in doing it at all. It's a useful thing to know if we can track it easily, but it looks like we can't in a meaningful way. Does it give us much more than we already know? The goal is to track Gutenberg adoption, and we can already track that, through posts published with more than a single classic block.

**Comment by creativecoder** [comment]
Note we regrouped on this at the end of last week: p1601667807280000-slack-CKZHG0QCR

It looks like there is a precedent from the previous Calypso editor for adding post meta like `_rest_api_published` and/or `_rest_api_client_id` for tracking what editor is used for a post. If we create this meta using hooks within the v2 REST API endpoints, this should give us what we need for knowing which editor was used.

**Comment by pablinos** [comment]
I've changed the approach slightly as I noticed that the classic editor plugin already stores the last editor used with a meta key of `classic-editor-remember`. I've copied this to [Jetpack with this PR](https://github.com/Automattic/jetpack/pull/17407), and then incorporated the changes on WPCOM with D50744-code.

It still needs some work, but it looks like this approach will work, and gets around the issue of other clients using the REST API etc.

**Comment by creativecoder** [comment]
Recording here what @pablinos and I discussed today:

First, we'll use the general approach we started on in Jetpack (https://github.com/Automattic/jetpack/pull/17407), but decouple it from the Classic editor plugin (use a different meta key, and one that's hidden). This way if we decide to track other types of clients as the "last used editor" (such as the editor in one of the mobile apps), it won't break the Classic editor plugin.

Down the road, we'd like to add something to the API endpoints for posts that allow specifying a client type that provide this meta value, but that will be handled separately, not in this issue.
